src/iterator/backend.rs

```diff
@@ -31,70 +32,69 @@ use crate::SigId;
 /// Maximal signal number we support.
 const MAX_SIGNUM: usize = 128;
 
-#[derive(Debug)]
-struct Waker<W> {
-    pending: Arc<[AtomicBool]>,
-    write: W,
+trait SelfPipeWrite: Debug {
+    fn wake_readers(&self);
 }
 
-impl<W: AsRawFd> Waker<W> {
-    /// Create a new instance for the given read and write pipe ends.
-    fn new(write: W) -> Self {
-        let pending: Arc<[AtomicBool]> = (0..MAX_SIGNUM)
-            .map(|_| AtomicBool::new(false))
-            .collect::<Vec<AtomicBool>>()
-            .into();
-
-        Self { pending, write }
+impl<W: AsRawFd + Debug> SelfPipeWrite for W {
+    fn wake_readers(&self) {
+        pipe::wake(self.as_raw_fd(), WakeMethod::Send);
     }
+}
 
-    /// Get a [`Pending`](struct.Pending) instance for iterating through recieved signals.
-    fn pending(&self) -> Pending {
-        Pending::new(Arc::clone(&self.pending))
-    }
+#[derive(Debug)]
+struct DeliveryState {
+    closed: AtomicBool,
+    registered_signal_ids: Mutex<Vec<Option<SigId>>>,
+}
 
-    /// Sends a wakeup signal to the internal wakeup pipe.
-    fn wake(&self) {
-        pipe::wake(self.write.as_raw_fd(), WakeMethod::Send);
+impl DeliveryState {
+    fn new() -> Self {
+        let ids = (0..MAX_SIGNUM).map(|_| None).collect();
+        Self {
+            closed: AtomicBool::new(false),
+            registered_signal_ids: Mutex::new(ids),
+        }
     }
 }
 
-#[derive(Debug)]
-struct RegisteredSignals(Mutex<Vec<Option<SigId>>>);
-
-impl Drop for RegisteredSignals {
+impl Drop for DeliveryState {
     fn drop(&mut self) {
-        let lock = self.0.lock().unwrap();
+        let lock = self.registered_signal_ids.lock().unwrap();
         for id in lock.iter().filter_map(|s| *s) {
             crate::unregister(id);
         }
     }
 }
 
-/// A struct to control an associated [`SignalDelivery`] or
-/// [`SignalIterator`] instance.
+/// A struct to control an instance of an associated type
+/// (like for example [`Signals`][super::Signals]).
 ///
 /// It allows to register more signal handlers and to shutdown the signal
-/// delivery. You may retrieve an instance of this type from the
-/// [`SignalDelivery::controller`] method.
-#[derive(Debug)]
-pub struct Controller<W> {
-    waker: Arc<Waker<W>>,
-    closed: AtomicBool,
-    registerd_signal_ids: RegisteredSignals,
+/// delivery. You can [`clone`][Handle::clone] this type which isn't a
+/// very expensive operation. The cloned instances can be shared between
+/// multiple threads.
+#[derive(Debug, Clone)]
+pub struct Handle {
+    pending: Arc<[AtomicBool]>,
```

Comment:
:thinking: Conceptually, the pending doesn't logically fit into the handle. If I read it correctly, it is used only to return the iterators. Should the `pending` live inside `SignalDelivery` directly? Or do I miss something?